x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetReader.java

```diff
@@ -152,21 +153,42 @@ public CacheHelper getReaderCacheHelper() {
         }
     }
 
-    private final BitSet roleQueryBits;
-    private final int numDocs;
+    private final DocumentSubsetBitsetCache bitsetCache;
+    private final Query roleQuery;
+    private BitSet roleQueryBits;
```

Comment:
does this one need to be volatile too?

Reply:
I thought it wouldn't be needed since we access the variable after the synchronized block ?

Reply:
What about access from getLiveDocs?

Reply:
We discussed offline and I added a comment in the code to explain why this is not needed (happens-before): 
https://github.com/elastic/elasticsearch/pull/48036/commits/9e98a69501b6963db17f396fb699386ed5a9b8f8